client/signup/steps/site-or-domain/index.jsx

```diff
@@ -18,25 +18,34 @@ import { externalRedirect } from 'lib/route/path';
 
 export default class SiteOrDomain extends Component {
 	componentWillMount() {
-		const { queryObject } = this.props;
-		let isValidDomain = queryObject && queryObject.new;
+		if ( ! this.getDomainName() ) {
+			// /domains domain search is an external application to calypso,
+			// therefor a full redirect required:
+			externalRedirect( '/domains' );
+		}
+	}
+
+	getDomainName() {
+		const { queryObject, step } = this.props;
 
-		if ( isValidDomain ) {
-			const domainParts = queryObject.new.split( '.' );
+		let domain, isValidDomain = false;
+
+		if ( queryObject && queryObject.new ) {
+			domain = queryObject.new;
+		} else if ( step && step.domainItem && step.domainItem.meta ) {
```

Comment:
Is it possible that `domainItem` is truthy but `meta` is not? If not, I think a presence check for `domainItem.meta` is misleading here, as a `domainItem` should always have a meta property ([function](https://github.com/Automattic/wp-calypso/blob/master/client/lib/cart-values/cart-items.js#L371)).